kafka_consumer/check.py

```diff
@@ -29,35 +40,161 @@ def __init__(self, name, init_config, agentConfig, instances=None):
+
+                # Group partitions by topic in order to construct the OffsetRequest
+                partitions_grouped_by_topic = defaultdict(list)
+                # partitions_for_broker returns all partitions for which this
+                # broker is leader. So any partitions that don't currently have
+                # leaders will be missed. Ignore as they'll be caught on next check run.
+                for topic, partition in kafka_client.cluster.partitions_for_broker(broker.nodeId):
+                    partitions_grouped_by_topic[topic].append(partition)
+
+                # Construct the OffsetRequest
+                timestamp = -1  # -1 for latest, -2 for earliest
+                max_offsets = 1
+                highwater_offsets_request = OffsetRequest[0](
+                    replica_id=-1,
+                    topics=[
+                        (topic, [
+                            (partition, timestamp, max_offsets) for partition in partitions])
+                        for topic, partitions in partitions_grouped_by_topic.iteritems()])
+
+                # Send each request as a blocking call to keep life simple.
+                # For large clusters, be faster to send all requests and then in a
+                # in a second loop process all responses. But requires bookeeping to
+                # make sure all returned.
+                future = kafka_client.send(broker.nodeId, highwater_offsets_request)
+                kafka_client.poll(future=future)
+                assert future.succeeded()  # safety net, typically only false if request was malformed.
+
+                highwater_offsets_response = future.value
+                for topic, partitions in highwater_offsets_response.topics:
+                    for partition, error_code, offsets in partitions:
+                        if error_code == 0:
+                            highwater_offsets[(topic, partition)] = offsets[0]
+                        # Valid error codes:
+                        # https://cwiki.apache.org/confluence/display/KAFKA/A+Guide+To+The+Kafka+Protocol#AGuideToTheKafkaProtocol-PossibleErrorCodes.2
+                        elif error_code == -1:
+                            self.log.exception("Kafka broker returned UNKNOWN "
+                                "(error_code -1) for topic: %s, partition: %s. "
+                                "This should never happen.",
+                                topic, partition)
+                        elif error_code == 3:
+                            self.log.warn("Kafka broker returned "
+                                "UNKNOWN_TOPIC_OR_PARTITION (error_code 3) for "
+                                "topic: %s, partition: %s. This should only "
+                                "happen if the topic is currently being deleted.",
+                                topic, partition)
+                        elif error_code == 6:
+                            self.log.warn("Kafka broker returned "
```

Comment:
Leader change should be uncommon for most clusters, as typically only happens when a broker is joing/leaving the cluster. Even then, just adding a new broker to a cluster can't trigger this state, it's only when the broker is part of the ISR list of some partition.